diff --git a/CMakeLists.txt b/CMakeLists.txt
index b73bf97aa..97615a996 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -27,6 +27,17 @@ ENDIF()
 
 PROJECT(h2o)
 
+SET(VERSION_MAJOR "2")
+SET(VERSION_MINOR "3")
+SET(VERSION_PATCH "0")
+SET(VERSION_PRERELEASE "-DEV")
+SET(VERSION "${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}${VERSION_PRERELEASE}")
+SET(LIBRARY_VERSION_MAJOR "0")
+SET(LIBRARY_VERSION_MINOR "16")
+SET(LIBRARY_VERSION_PATCH "0")
+SET(LIBRARY_VERSION "${LIBRARY_VERSION_MAJOR}.${LIBRARY_VERSION_MINOR}.${LIBRARY_VERSION_PATCH}${VERSION_PRERELEASE}")
+SET(LIBRARY_SOVERSION "${LIBRARY_VERSION_MAJOR}.${LIBRARY_VERSION_MINOR}")
+
 INCLUDE(GNUInstallDirs)
 INCLUDE(CheckCSourceCompiles)
 INCLUDE(CMakePushCheckState)
@@ -35,6 +46,7 @@ INCLUDE(deps/picotls/cmake/boringssl-adjust.cmake)
 INCLUDE(deps/picotls/cmake/dtrace-utils.cmake)
 INCLUDE(deps/picotls/cmake/fusion.cmake)
 
+CONFIGURE_FILE(${CMAKE_CURRENT_SOURCE_DIR}/version.h.in ${CMAKE_CURRENT_SOURCE_DIR}/include/h2o/version.h)
 CONFIGURE_FILE(${CMAKE_CURRENT_SOURCE_DIR}/libh2o.pc.in ${CMAKE_CURRENT_BINARY_DIR}/libh2o.pc @ONLY)
 CONFIGURE_FILE(${CMAKE_CURRENT_SOURCE_DIR}/libh2o-evloop.pc.in ${CMAKE_CURRENT_BINARY_DIR}/libh2o-evloop.pc @ONLY)
 
@@ -608,20 +620,6 @@ ELSE ()
     ADD_LIBRARY(libh2o-evloop ${LIB_SOURCE_FILES})
 ENDIF (WSLAY_FOUND)
 
-EXECUTE_PROCESS(
-    COMMAND sed -n -E "/#define {1,}H2O_VERSION_BASE .*-DEV\"$/s//YES/p" ${CMAKE_CURRENT_SOURCE_DIR}/include/h2o/version.h
-    OUTPUT_VARIABLE IS_PRERELEASE
-    OUTPUT_STRIP_TRAILING_WHITESPACE)
-EXECUTE_PROCESS(
-    COMMAND sed -n -E "/#define {1,}H2O_LIBRARY_VERSION {1,}\"(.*)\"$/s//\\1/p" ${CMAKE_CURRENT_SOURCE_DIR}/include/h2o/version.h
-    OUTPUT_VARIABLE LIBRARY_VERSION
-    OUTPUT_STRIP_TRAILING_WHITESPACE)
-EXECUTE_PROCESS(
-    COMMAND sed -n -E "/#define {1,}H2O_LIBRARY_VERSION {1,}\"(.*)\\.[0-9]{1,}\"$/s//\\1/p" ${CMAKE_CURRENT_SOURCE_DIR}/include/h2o/version.h
-    OUTPUT_VARIABLE LIBRARY_SOVERSION
-    OUTPUT_STRIP_TRAILING_WHITESPACE)
-MESSAGE(STATUS "extracted version numbers from ${CMAKE_CURRENT_SOURCE_DIR}/include/h2o/version.h:\n\tprerelease=${IS_PRERELEASE}\n\tlibrary version=${LIBRARY_VERSION}\n\tlibrary soversion=${LIBRARY_SOVERSION}")
-
 SET_TARGET_PROPERTIES(libh2o PROPERTIES
     OUTPUT_NAME h2o
     VERSION ${LIBRARY_VERSION}
diff --git a/include/h2o/version.h b/include/h2o/version.h
index c22a22a24..706d3112a 100644
--- a/include/h2o/version.h
+++ b/include/h2o/version.h
@@ -31,13 +31,6 @@
 #define H2O_VERSION_MINOR 3
 #define H2O_VERSION_PATCH 0
 
-#ifdef H2O_GITREV
-#define H2O_VERSION H2O_VERSION_BASE "@" H2O_TO_STR(H2O_GITREV)
-#else
-#define H2O_VERSION H2O_VERSION_BASE
-#endif
-
-/* `H2O_LIBRARY_VERSION` is a hard-coded string with three digits, that's the format we parse in CMakeLists.txt */
 #define H2O_LIBRARY_VERSION "0.16.0"
 #define H2O_LIBRARY_VERSION_MAJOR 0
 #define H2O_LIBRARY_VERSION_MINOR 16
diff --git a/version.h.in b/version.h.in
new file mode 100644
index 000000000..fc6db3d1f
--- /dev/null
+++ b/version.h.in
@@ -0,0 +1,22 @@
+#ifndef h2o__version_h
+#define h2o__version_h
+
+#ifdef H2O_HAS_GITREV_H
+#include "h2o/gitrev.h"
+#endif
+
+#ifdef H2O_GITREV
+#define H2O_VERSION "@VERSION@@" H2O_TO_STR(H2O_GITREV)
+#else
+#define H2O_VERSION "@VERSION@"
+#endif
+
+#define H2O_VERSION_MAJOR @VERSION_MAJOR@
+#define H2O_VERSION_MINOR @VERSION_MINOR@
+#define H2O_VERSION_PATCH @VERSION_PATCH@
+
+#define H2O_LIBRARY_VERSION_MAJOR @LIBRARY_VERSION_MAJOR@
+#define H2O_LIBRARY_VERSION_MINOR @LIBRARY_VERSION_MINOR@
+#define H2O_LIBRARY_VERSION_PATCH @LIBRARY_VERSION_PATCH@
+
+#endif
